src/lib.rs.in

```diff
@@ -47,87 +46,103 @@ fn json_response(obj: &Value) -> IronResult<Response> {
 }
 
 
-/// Configuration data for a "famous" identity provider.
-///
-/// Used as part of the `AppConfig` struct.
+/// Holds runtime configuration data for this daemon instance.
 #[derive(Clone)]
-struct ProviderConfig {
-    /// URL pointing to the OpenID configuration document.
+pub struct AppConfig {
+    base_url: String, // Origin of this instance, used for constructing URLs
+    priv_key: PKey, // Signing key
+    store: redis::Client, // Redis Client
+    expire_keys: usize, // Redis key TTL, in seconds
+    sender: Email, // From address for email
+    token_validity: usize, // JWT validity duration, in seconds
+    providers: HashMap<String, Provider>, // Mapping of Domain -> OIDC Provider
+}
+
+
+/// Represents an OpenID Connect provider.
+#[derive(Clone, Deserialize)]
+pub struct Provider {
     discovery: String,
-    /// Client ID issued for this daemon instance by the provider.
     client_id: String,
-    /// Secret issued for this daemon instance by the provider.
     secret: String,
-    /// Issuer origin as used in identity tokens issued by the provider.
-    /// Used to check that the issuer in the token matches expectations.
-    issuer: String,
 }
 
 
-/// Configuration data for this daemon instance.
-#[derive(Clone)]
-pub struct AppConfig {
-    /// Version of the daemon (used in the `WelcomeHandler`).
-    version: String,
-    /// Base URL where this daemon can be found. This is used to construct
-    /// callback URLs.
+/// Represents an email address.
+#[derive(Clone, Deserialize)]
+pub struct Email {
+    address: String,
+    name: String,
+}
+
+
+/// Represents a JSON-based configuration file.
+#[derive(Clone, Deserialize)]
+pub struct JsonConfig {
     base_url: String,
-    /// Private key used to sign identity tokens.
-    priv_key: PKey,
-    /// Redis database connector.
-    store: Client,
-    /// Duration in seconds for expiry of data stored in Redis.
+    private_key_file: String,
+    redis_url: String,
+    sender: Email,
     expire_keys: usize,
-    /// Email sender (email address, then human-readable name).
-    sender: (String, String),
-    /// Duration in seconds for expiration of identity tokens.
-    token_validity: i64,
-    /// A map of "famous" identity providers. Each key is an email domain,
-    /// the value holds the configuration required to act as an application
-    /// doing OpenID Connect against the provider.
-    providers: BTreeMap<String, ProviderConfig>,
+    token_validity: usize,
+    providers: HashMap<String, Provider>,
+}
+
+
+/// Union of all possible error types seen while parsing.
+#[derive(Debug)]
+pub enum Error {
```

Comment:
If you can find one, I'm all ears ;) Maybe a macro, but that might be less obvious.